File: backend/app/models/notification.py
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.alert import Alert
    from app.models.incident import Incident
    from app.models.user import User


class Notification(Base):
    """In-app notification delivered to a single user."""

    __tablename__ = "notifications"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    type: Mapped[str] = mapped_column(String(50), nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    is_read: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        server_default="false",
        nullable=False,
    )
    incident_id: Mapped[int | None] = mapped_column(
        ForeignKey("incidents.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    alert_id: Mapped[int | None] = mapped_column(
        ForeignKey("alerts.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    user: Mapped["User"] = relationship(back_populates="notifications")
    incident: Mapped["Incident | None"] = relationship()
    alert: Mapped["Alert | None"] = relationship()

File: frontend/src/lib/mappers/notification.ts
import type { ApiNotification } from '../../types/api/notification.ts'
import type { Notification } from '../../types/notification.ts'
import { parseIncidentNumericId } from '../incident-utils.ts'

export function mapApiNotificationToNotification(api: ApiNotification): Notification {
  return {
    id: String(api.id),
    type: api.type,
    title: api.title,
    message: api.message,
    isRead: api.is_read,
    incidentId: api.incident_id !== null ? `inc-${api.incident_id}` : undefined,
    alertId: api.alert_id !== null ? String(api.alert_id) : undefined,
    createdAt: api.created_at,
  }
}

export function getNotificationIncidentPath(notification: Notification): string | null {
  if (!notification.incidentId) {
    return null
  }

  const incidentId = parseIncidentNumericId(notification.incidentId)
  return incidentId ? `/incidents/${incidentId}` : null
}
